packages/flutter_tools/bin/xcode_backend.dart

```diff
@@ -188,55 +238,131 @@ class Context {
     // which could be a local build or an arch/type specific build.
-    runRsync(
-      delete: true,
-      '${environment['BUILT_PRODUCTS_DIR']}/Flutter.framework',
-      '$xcodeFrameworksDir/',
-    );
+    switch (platform) {
+      case TargetPlatform.ios:
+        runRsync('${environment['BUILT_PRODUCTS_DIR']}/Flutter.framework', '$xcodeFrameworksDir/');
+      case TargetPlatform.macos:
+        runRsync(
+          extraArgs: <String>['--filter', '- Headers', '--filter', '- Modules'],
+          '${environment['BUILT_PRODUCTS_DIR']}/FlutterMacOS.framework',
+          '$xcodeFrameworksDir/',
+        );
+
+        if (codesign) {
+          _codesignFramework(expandedCodeSignIdentity, '$xcodeFrameworksDir/App.framework/App');
+          _codesignFramework(
+            expandedCodeSignIdentity,
+            '$xcodeFrameworksDir/FlutterMacOS.framework/FlutterMacOS',
+          );
+        }
+    }
 
-    // Copy the native assets. These do not have to be codesigned here because,
-    // they are already codesigned in buildNativeAssetsMacOS.
+    // Copy the native assets.
     final String sourceRoot = environment['SOURCE_ROOT'] ?? '';
     String projectPath = '$sourceRoot/..';
     if (environment['FLUTTER_APPLICATION_PATH'] != null) {
       projectPath = environment['FLUTTER_APPLICATION_PATH']!;
     }
     final String flutterBuildDir = environment['FLUTTER_BUILD_DIR']!;
-    final String nativeAssetsPath = '$projectPath/$flutterBuildDir/native_assets/ios/';
+    final String nativeAssetsPath = '$projectPath/$flutterBuildDir/native_assets/${platform.name}/';
     final bool verbose = (environment['VERBOSE_SCRIPT_LOGGING'] ?? '').isNotEmpty;
-    if (Directory(nativeAssetsPath).existsSync()) {
+    final Directory nativeAssetsDir = directoryFromPath(nativeAssetsPath);
+    if (nativeAssetsDir.existsSync()) {
       if (verbose) {
         print('♦ Copying native assets from $nativeAssetsPath.');
       }
-      runRsync(
-        extraArgs: <String>['--filter', '- native_assets.yaml', '--filter', '- native_assets.json'],
-        nativeAssetsPath,
-        xcodeFrameworksDir,
-      );
+      for (final FileSystemEntity entity in nativeAssetsDir.listSync()) {
+        if (entity is Directory) {
+          final String? frameworkName = parseFrameworkNameFromDirectory(entity);
+          if (frameworkName != null) {
+            runRsync(
+              extraArgs: <String>[
+                '--filter',
+                '- native_assets.yaml',
+                '--filter',
+                '- native_assets.json',
+              ],
+              entity.path,
+              xcodeFrameworksDir,
+            );
+            if (codesign) {
+              _codesignFramework(
+                expandedCodeSignIdentity,
+                '$xcodeFrameworksDir/$frameworkName.framework/$frameworkName',
+              );
+            }
+          }
+        }
+      }
     } else if (verbose) {
```

Comment:
I'd consider moving the native assets logic into a helper function since it's gotten quite big, and then switching this to be an early return